### ISSUE TITLE
2 issues with the virtual 'service' service and service.py in particular

### ISSUE THREAD

**Issue by cr1st1p** [saltstack/salt#49465]
### Description of Issue/Question
Virtual module 'service' behaving wrong on a Manjaro distro - but the issues do NOT seem to be particular to Manjaro.

Issues:

-  a) saltstack seem to not load one and **only** one service implementation module
-  b) has kind of wrong conditions for using the 'default' _service.py_ module

I'll propose a small change that would fix b) although tests should be run on the affected platforms.

### Setup
When trying to force the service module, I am adding a 'providers.conf' file in /etc/salt/minion.d/ :
```
providers:
   service: systemd

```
### Steps to Reproduce Issue
On a Manjaro distro, freshly installed (KDE, 17.1.12), install 'salt' via pacman (`pacman -Su salt`)
Then run
`salt-call --local --refresh-grains-cache -l debug service.get_running`
You will get a list of running service. Among them, 'sddm' let's say.

Let's get the full list:
`salt-call --local --refresh-grains-cache -l debug service.get_all`
Surprise! List is empty.

Let's check if 'sddm' service is available:
`salt-call --local --refresh-grains-cache -l debug service.available sddm`
Result: False (!!)

Let's edit 'service.py' (/usr/lib/python2.7/site-packages/salt/modules/service.py) and rename function `get_available` to `get_available2` and run the command again.
Result: True (!!) 
So... service.py (using /etc/init.d/ files) gets **ALSO** loaded together with systemd.py one and functions are merged from both modules !?
This looks like a wrong implementation of the concept: I have one and only one service module implementation.
Or, unless by design, it is ok to maybe load multiple python modules for a single virtual service.

Renaming `service.py:get_all()` to `service.py:get_all2()` will also show good results for the _service.get_all_ salt-call command run.

If I add the providers.conf file and ensure it is listed in the debug output of salt-call run, things seem to work.

One suggested patch:
```
--- /root/service.py.orig	2018-08-31 11:47:59.689732154 +0300
+++ /root/service.py	2018-08-31 11:39:55.767772241 +0300
@@ -9,6 +9,8 @@
 import fnmatch
 import re
 
+import salt.utils.systemd
+
 __func_alias__ = {
     'reload_': 'reload'
 }
@@ -50,8 +52,12 @@
         'XenServer',
         'Cumulus'
     ))
-    if __grains__.get('os', '') in disable:
+    if __grains__.get('os', '') in disable or __grains__.get('os_family', '') in disable:
         return (False, 'Your OS is on the disabled list')
+    if __grains__['kernel'] == 'Linux' \
+            and salt.utils.systemd.booted(__context__):
+	return (False, 'Your OS is using systemd')
+    
     # Disable on all non-Linux OSes as well
     if __grains__['kernel'] != 'Linux':
         return (False, 'Non Linux OSes are not supported')

```
The first condition change was added in the idea that distros based on some more general ones, will behave the same, so that's why I added the check on 'os_family'.
Note that 'Manjaro' is a derivative of 'Arch' distro, and it shows with os_family as 'Arch'.
This needs some tests, at least for Suse distro(s).

The second change would ensure that once _systemd_ use is detected then it would not use the current module. This is a more generic condition, until a proper fix for using a single service module implementation is done.

### Versions Report
```
Salt Version:
           Salt: 2018.3.2
 
Dependency Versions:
           cffi: 1.11.5
       cherrypy: Not Installed
       dateutil: Not Installed
      docker-py: Not Installed
          gitdb: Not Installed
      gitpython: Not Installed
          ioflo: Not Installed
         Jinja2: 2.10
        libgit2: 0.27.0
        libnacl: Not Installed
       M2Crypto: 0.30.1
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.5.6
   mysql-python: Not Installed
      pycparser: 2.18
       pycrypto: 2.6.1
   pycryptodome: Not Installed
         pygit2: 0.27.1
         Python: 2.7.15 (default, Jun 27 2018, 13:05:28)
   python-gnupg: Not Installed
         PyYAML: 3.13
          PyZMQ: 17.1.0
           RAET: Not Installed
          smmap: Not Installed
        timelib: Not Installed
        Tornado: 5.1
            ZMQ: 4.2.5
 
System Versions:
           dist:   
         locale: UTF-8
        machine: x86_64
        release: 4.19.0-1-MANJARO
         system: Linux
        version: Not Installed

```

**Comment by Ch3LL** [comment]
im having a hard time replicating this. If i dont set the provider settings it calls `service.py` and if i set `systemd` as you have in my minion config it only calls `get_all` in systemd file. I added debug statements and it never hits service.py's get_all method. 

did you recently upgrade salt possibly?
can you add htis line `log.error(__opts__['providers'])` in the get_all function that is getting called and make sure the providers is being set correctly as you would expect.

If manjaro is indeed not suppose to ever run service.py the fix will most likely be:

```
diff --git a/salt/modules/service.py b/salt/modules/service.py
index a041fb1e0f..b8ce0bfd4a 100644
--- a/salt/modules/service.py
+++ b/salt/modules/service.py
@@ -44,6 +44,7 @@ def __virtual__():
         'Linaro',
         'elementary OS',
         'McAfee  OS Server',
+        'Manjaro',
         'Void',
         'Mint',
         'Raspbian',
```

can you test that fix

**Comment by cr1st1p** [comment]
To replicate, do NOT set the providers setting. Then call get_running, get_all and available.

While your fix will probably fix strictly my problem, it is the simplest path possible. The more general issue of having two modules mixed still remains.

Manjaro and Arch distros also supports openrc style services, so I really would keep at least the systemd check that I added (my 2nd 'if')
Or just fix code to load only one module.

**Comment by Ch3LL** [comment]
if i set or do not set providers I cannot replicate `get_all` trying to load two modules. It only loads one module.

**Comment by cr1st1p** [comment]
First - do not edit and use any provider setting. Let salt use whatever it detects.

Next, keep in mind that file service.py is for sysvinit type of systems and systemd.py file for... systemd type of systems.
Usually, these type of systems are mutually exclusive (but not always). 

Based on saltstack's implementation, especially of service.py file - these python modules should be mutually exclusive.
See also this comment in service.py: " Only work on systems which exclusively use sysvinit".

Next: my system, Manjaroo, is a *systemd* type system. As such, obviously, there should be NO functions called from file service.py.

It looks like until now you called ONLY service.get_all function and none of the other 2. And you discovered that the function is called from file service.py - which is wrong
since my system is systemd type and not 'exclusively use sysvinit'. And of course it gives totally wrong results - i.e. an empty list!

If you run *ALL* these 3 calls (service.get_all, service.get_running, service.available), you will see that the actual function is called from
either service.py or systemd.py files, and ... these should be mutually exclusive! One and only one file should have all those functions called.


Let's try also a table:


On my distro (Manjaro), systemd style:

```
Salt Module function call	| code run from service.py	| code run from systemd.py	|
			    	| ("exclusiveley use sysvinit")	|  (systemd type systems) 	|
--------------------------------+-------------------------------+-------------------------------|
service.get_all()		|      	yes (WRONG!)		|            	-		|
service.get_running()		|            - 			|    yes (correct)		|
service.available('something')	| 	yes (WRONG!)		|       	-		|
-------------------------------------------------------------------------------------------------

```

**Comment by Ch3LL** [comment]
my understanding is that either your fix or my fix needs to be added to fix the issue. like you said you fix will add some more logic so we can just see if systemd is loaded.

The calls to `get_all` and `available` are called on service.py because that is whats loaded first. `get_running` is getting loaded from systemd because `get_running` does not exist in service.py.

But i want to get @saltstack/team-core 's opinons here to make sure i am understanding this correctly?

**Comment by terminalmage** [comment]
This will be fixed in 2018.3.3, see https://github.com/saltstack/salt/pull/48955.

**Comment by cr1st1p** [comment]
checked today on salt version 2019.2.2 and it seems to work ok. Code also looks fine.
closing. 
Thank you for the fix!
